README.md
Description
===========

* Installs OSEv3 and related packages.

Requirements
============

## Platform ##

* Tested on Red Hat RHEL 7.2
* Tested on Centos 7.2

## Openshift Version ##

* Support OSE version from 3.1.0.4+
* Support Origin version from 1.1.1+
* Default the installation to 1.4 or 3.4

**Highly recommended**: explicitly set `node['cookbook-openshift3']['ose_version']`, `node['cookbook-openshift3']['ose_major_version']`
and ideally `node['cookbook-openshift3']['docker_version']` to be safe when a major version is released on the
CentOS PaaS repository; this cookbook does NOT support upgrade between major versions, so lock your package versions
in your openshift3 role or environment.

Test Matrix
===========

| Platform | OSE 1.4.1 | OSE 1.3.3 | OSE 1.2.1 |
|----------|-----------|-----------|-----------|
| centos 7.2 | PASS | PASS | open issues: #76 |

Override Attributes
===================

[Read more about overriding attributes here!](https://github.com/IshentRas/cookbook-openshift3/blob/master/attribute-cookbook.md)

#### Structure ####
-------------------

* `node['cookbook-openshift3']['openshift_master_identity_provider']['HTPasswdPasswordIdentityProvider']`

```json
{
"name" : "htpasswd_auth", 
"login" : true, 
"challenge" : true, 
"kind" : "HTPasswdPasswordIdentityProvider", 
"filename" : "/etc/openshift/openshift-passwd"
}
```

* `node['cookbook-openshift3']['openshift_master_identity_provider']['LDAPPasswordIdentityProvider']`

```json
{
"name" : "ldap_identity", 
"login" : true, 
"challenge" : true, 
"kind" : "LDAPPasswordIdentityProvider", 
"ldap_server" : "ldap.domain.local",
 "ldap_bind_dn" : "", 
 "ldap_bind_password" : "",
 "ldap_insecure" : true, 
 "ldap_base_ou" : "OU=people,DC=domain,DC=local", 
 "ldap_preferred_username" : "uid"
}
```

* `node['cookbook-openshift3']['openshift_master_asset_config']`

```json
{
 "extensionStylesheets":["/path/to/css"],
 "extensionScripts":["/path/to/script"],
 "extensions":"/path/to/my_images",
 "templates":"/path/to/template"
}
```

* `node['cookbook-openshift3']['openshift_node_docker-storage']`

```json
{
 "DEVS": ["/dev/sdb","/dev/sdc"],
 "VG": "docker-vg",
 "DATA_SIZE": "80%FREE"
}
```

* `node['cookbook-openshift3']['docker_log_driver']`

Set to `'json-file'` (default), `'journald'` or any other supported [docker log driver](https://docs.docker.com/engine/admin/logging/overview/).
Set to '' to disable it.

* `node['cookbook-openshift3']['docker_log_options']`

Assuming `node['cookbook-openshift3']['docker_log_driver']` is `'json-file'` (the default):

```json
{
 "max-size": "50M",
 "max-file": "3"
}
```

Any option can be set, as long as they are supported by the current [docker log driver](https://docs.docker.com/engine/admin/logging/overview/).


* `node['cookbook-openshift3']['openshift_master_named_certificates']`

* CN or SAN names are automatically detected from the certificate file.

```json
[
  {
   "certfile": "/etc/where_is/my_certfile",
   "keyfile": "/etc/where_is/my_keyfile"
  }
]
```

* `node['cookbook-openshift3']['persistent_storage']`

* They key called 'claim' is optional and will automatically create a PersistentVolumeClaim within a specified namespace.
 
```json
{
  "name": "registry",
  "capacity": "5Gi",
  "access_modes": "ReadWriteMany",
  "server": "core.domain.local",
  "path": "/var/nfs/registry",
  "claim": {
    "namespace": "default"
  }
}
```
* `node['cookbook-openshift3']['openshift_hosted_metrics_parameters']`

Any option can be set, as long as they are supported by the current [Metrics deployer template](https://docs.openshift.com/container-platform/latest/install_config/cluster_metrics.html#deployer-template-parameters).

The name of the key can be set in upper case or lower case

```json
{
  "openshift_master_metrics_public_url": "metrics.domain.local",
  "openshift_hosted_metrics_parameters": {
    "HAWKULAR_METRICS_HOSTNAME": "metric.domain.local",
    "METRIC_DURATION": "30",
    "IMAGE_VERSION": "v1.4.1"
  }
}
```

Remember to specify an `IMAGE_VERSION` or the `latest` version will be deployed, which may be incompatible with your openshift cluster version.

## Cloud Providers Integration

Cloud providers integration requires passing some sensetive credentials to OpenShift. This cookbook uses encrypted data bags as the safest way to achieve this. Thus you should have: 

- running Chef Server
- encrypted data bags with cloud providers' credentials

*Currently only AWS integration is supported.*

### AWS 

To integrate your OpenShift installation with AWS you should have following attributes for `cookbook-openshift3` cookbook:

```json
{
  "openshift_cloud_provider": "aws",
  "openshift_cloud_providers": {
    "aws": {
      "data_bag_name": "cloud-provider",
      "data_bag_item_name": "aws",
      "secret_file": "/etc/chef/shared_secret"
    }
  }
}
```

You should also have data bag named `cloud-provider` (`data_bag_name` attribute above) and encrypted with some shared secret data bag item named `aws` (`data_bag_item_name` attribute above) at your Chef Server. If `secret_file` attribute from above is *not* defined a default for Chef Client shared secret file will be used (`/etc/chef/encrypted_data_bag_secret`). For more information see [official Chef docs](https://docs.chef.io/data_bags.html#encrypt-a-data-bag-item).

Data bag item content should be of the form:

```json
{
  "id": "aws",
  "access_key_id": "your_access_key_id",
  "secret_access_key": "your_secret_access_key"
}
```

Please note: `id` value should be exactly the same as `data_bag_item_name` attribute value from above.

**Alternatively** you can attach IAM policies to your AWS instances and do *not* provide AWS credentials in encrypted data bags. In this case you should have the following attribute for `cookbook-openshift3` cookbook:

```json
{
  "openshift_cloud_provider": "aws"
}
```

and the following IAM policy attached to your *master* servers:

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": "ec2:*",
      "Resource": "*"
    },
    {
      "Effect": "Allow",
      "Action": "elasticloadbalancing:*",
      "Resource": "*"
    }
  ]
}
```

and the following IAM policy attached to your *node* servers: 

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": "ec2:Describe*",
      "Resource": "*"
    },
    {
      "Effect": "Allow",
      "Action": "ec2:AttachVolume",
      "Resource": "*"
    },
    {
      "Effect": "Allow",
      "Action": "ec2:DetachVolume",
      "Resource": "*"
    }
  ]
}
```

=====

Include the default recipe in a CHEF role so as to ease the deployment. 

## Roles (Examples) 

* BASE

```json
{
  "name": "base",
  "description": "Common Base Role",
  "json_class": "Chef::Role",
  "default_attributes": {

  },
  "override_attributes": {
  },
  "chef_type": "role",
  "run_list": [
    "recipe[cookbook-openshift3]"
  ],
  "env_run_lists": {

  }
}
```

* UNINSTALL (ADHOC)

```json
{
  "name": "uninstall",
  "description": "Common Base Role",
  "json_class": "Chef::Role",
  "default_attributes": {

  },
  "override_attributes": {
  },
  "chef_type": "role",
  "run_list": [
    "recipe[cookbook-openshift3::adhoc_uninstall]"
  ],
  "env_run_lists": {

  }
}
```

ENVIRONMENT
===========

Modify the attributes as required in your environments to change how various configurations are applied per the attributes section above. 
In general, override attributes in the environment should be used when changing attributes.

### Minimal example ###

* CLUSTER-NATIVE (Only available option)

```json
{
  "name": "cluster_native",
  "description": "",
  "cookbook_versions": {

  },
  "json_class": "Chef::Environment",
  "chef_type": "environment",
  "default_attributes": {

  },
  "override_attributes": {
    "cookbook-openshift3": {
      "openshift_HA": true,
      "openshift_cluster_name": "ose-cluster.domain.local",
      "persistent_storage": [
        {
          "name": "registry",
          "capacity": "5Gi",
          "access_modes": "ReadWriteMany",
          "server": "core.domain.local",
          "path": "/var/nfs/registry",
          "claim": {
            "namespace": "default"
          }
        },
        {
          "name": "metrics",
          "capacity": "3Gi",
          "access_modes": "ReadWriteOnce",
          "server": "core.domain.local",
          "path": "/var/nfs/metrics",
          "policy": "Recycle"
        },
        {
          "name": "logging",
          "capacity": "2Gi",
          "access_modes": "ReadWriteOnce",
          "server": "core.domain.local",
          "path": "/var/nfs/logging",
          "policy": "Recycle"
        }
      ],
      "master_servers": [
        {
          "fqdn": "ose1-server.domain.local",
          "ipaddress": "1.1.1.1"
        },
        {
          "fqdn": "ose2-server.domain.local",
          "ipaddress": "1.1.1.2"
        },
        {
          "fqdn": "ose3-server.domain.local",
          "ipaddress": "1.1.1.3"
        }
      ],
      "etcd_servers": [
        {
          "fqdn": "ose1-server.domain.local",
          "ipaddress": "1.1.1.1"
        },
        {
          "fqdn": "ose2-server.domain.local",
          "ipaddress": "1.1.1.2"
         
        },
        {
          "fqdn": "ose3-server.domain.local",
          "ipaddress": "1.1.1.3"
        }
      ],
      "node_servers": [
        {
          "fqdn": "ose1-server.domain.local",
          "ipaddress": "1.1.1.1",
          "schedulable": true,
          "labels": "region=infra"
        },
        {
          "fqdn": "ose2-server.domain.local",
          "ipaddress": "1.1.1.2",
          "schedulable": true,
          "labels": "region=infra"
        },
        {
          "fqdn": "ose3-server.domain.local",
          "ipaddress": "1.1.1.3",
          "schedulable": true,
          "labels": "region=infra"
        },
        {
          "fqdn": "ose4-server.domain.local",
          "ipaddress": "1.1.1.4",
          "labels": "region=user zone=east"
        },
        {
          "fqdn": "ose5-server.domain.local",
          "ipaddress": "1.1.1.5",
          "labels": "region=user zone=west"
        },        
      ],
    }
  }
}
```
* SINGLE MASTER (EMBEDDED ETCD)

```json
{
  "name": "single",
  "description": "",
  "cookbook_versions": {

  },
  "json_class": "Chef::Environment",
  "chef_type": "environment",

CHANGELOG.md
# Openshift 3 Cookbook CHANGELOG
This file is used to list changes made in each version of the Openshift 3 cookbook.

## v1.10.46
### Bug
- Spacing corrected and defaults not included as per discussion in #115

## v1.10.45
### Improvement
- Jenkinsfile has resilient kitchen tests reinstated

### Bug
- Certificate redeployment code fixed to remove node certs
- Certificate redeployment - run etcd code only if etcd on the node (eg standalone)

## v1.10.44
### Improvement
- Upgrade from x.2 to x.3 supported
- Service signer cert created as part of cert creation

## v1.10.43
### Improvement
- Added Jenkinsfile
- Rename file: service_openvsitch-containerized.service.erb -> service_openvswitch-containerized.service.erb

### Bug
- Fix redeploy certs for separate etcd cluster
- Use more config items rather than hard-coded values in delete node

## v1.10.42
### Improvement
- Replace most hard link usage with local copy
- Change http server binding to default IP address
- Give the possibility to specify custom certificate for hosted router

### Bug
- Fix CHEF-3694 warning due to redundant package resource
- Removed potentially unsafe identity providers defaults
- Fix named certificates when common_name is also listed in alternative names
- Allow distinct hostnames for internal and public API access

## v1.10.41
### Bug
- Fix AWS issue when using empty data bag

## v1.10.40
### Bug
- Fix bug for enterprise version (Hosted template files)
- Fix deletion of service files

## v1.10.39
### Improvement
- Give the possibility of adding custom master CA certificate
- Give the possibility of supporting AWS IAM based integration

### Bug
- Fix dnsIP for dedicated nameserver within PODS (Default to IP of the node)

## v1.10.38
### Improvement
- Give the possibility of adding cAdvisor port and read-only port for kubelet arguments
- Give the possibility of skipping nodes when applying schedulability and labelling

### Bug
- Fix issue reported by https://github.com/IshentRas/cookbook-openshift3/issues/77

## v1.10.37
### Bug
- Emergency update for fixing ose_major_version when running standalone deployment

## v1.10.36
### Bug
- Fixed cookstyle offenses
- Revert the ETCD change causing issue when adding / removing members
- Make secret call compatible with x.2 version(s)

### Improvement
- Expand .kitchen.yml to test OSE v1.4.1, v1.3.3 and v1.2.1
- Improved code readability
- Added support for multiple identity providers
- Added support for AWS cloud provider

## v1.10.35
### Bug
- Fix ETCD service defined in 2 places
- Fix cookstyle issues
- Fix admin.kubeconfig logic

## v1.10.34
### Bug
- Adjust predicates and priorities based on ose_major_version
- Fix containerized deployment

### Improvement
- Give the possibility to add or remove etcd server members
- Improve ETCD deployment for single etcd server

## v1.10.33
### Bug
- Revert e168f9b, use stable repository URLs again

### Improvement
- Use stable CentOS PaaS repository during tests
- Add integration test for hosted metrics feature

## v1.10.32
### Bug
- Make apiServerArguments conditional on the version for pre-1.3/3.3 versions

## v1.10.31
### Improvement
- Handle 1.4/3.4 deployment
- Clean codes over unused attributes
- Integration tests for 1.4/3.4
- Add the possibility to supply dns-search option via Docker
- Add the possibility to specify a deserialization cache size parameter. 

### Bug
- Fix permissions over /etc/origin/node
- Fix iptables issue due to version used by clients

## v1.10.30
### Improvement
- Add the possibility to deploy the cluster metrics
- Add the possibility to add more manageName serviceaccount in master config
- Move registry persistent_volume_claim name to explicit LWRP attribute
- Added integration test for openshift_hosted_manage_registry feature
- Added integration test for openshift_hosted_manage_router feature
- Added integration test for persistent_storage feature
- Refactor router-related resources to new openshift_deploy_router LWRP
- Move registry persistent_volume_claim name to explicit LWRP attribute

### Bug
- Fix README.md typo
- Fix issue with systemd when uninstalling the Openshift
- Fix issue for systemctl daemon-reload
- Removed redundant guard clause for registry deloyment

## v1.10.29
### Bug
- Remove property attributes for resources (backward compatibility)

## v1.10.28
### Improvement
- Add the possibility to deploy the cluster metrics
- Add the possibility to add more manageName serviceaccount in master config

### Bug
- Fix README.md typo
- Fix issue with systemd when uninstalling the Openshift
- Fix issue for systemctl daemon-reload

## v1.10.27
### Bug
- Fix the Origin deployment issue (https://github.com/IshentRas/cookbook-openshift3/issues/20)
- Fix master-api service and master-controllers service (https://github.com/IshentRas/cookbook-openshift3/issues/40)

## v1.10.26
### Improvement
- Set the default ipaddress used in etcd-related attributes accordingly with the etcd_server variable

### Bug
- Remove duplicated variables for ETCD

## v1.10.25
### Bug
- Fix documentation
- Fix redeploying OSE certificates

## v1.10.24
### Improvement

attributes/default.rb
#
# Cookbook Name:: cookbook-openshift3
# Recipe:: default
#
# Copyright (c) 2015 The Authors, All Rights Reserved.

default['cookbook-openshift3']['use_wildcard_nodes'] = false
default['cookbook-openshift3']['wildcard_domain'] = ''
default['cookbook-openshift3']['openshift_cluster_name'] = nil
default['cookbook-openshift3']['openshift_HA'] = false
default['cookbook-openshift3']['master_servers'] = []
default['cookbook-openshift3']['etcd_servers'] = []
default['cookbook-openshift3']['node_servers'] = []

if node['cookbook-openshift3']['openshift_HA']
  default['cookbook-openshift3']['openshift_common_api_hostname'] = node['cookbook-openshift3']['openshift_cluster_name']
  default['cookbook-openshift3']['openshift_common_public_hostname'] = node['cookbook-openshift3']['openshift_common_api_hostname']
  default['cookbook-openshift3']['openshift_master_embedded_etcd'] = false
  default['cookbook-openshift3']['openshift_master_etcd_port'] = '2379'
  default['cookbook-openshift3']['master_etcd_cert_prefix'] = 'master.etcd-'
else
  default['cookbook-openshift3']['openshift_common_api_hostname'] = node['fqdn']
  default['cookbook-openshift3']['openshift_common_public_hostname'] = node['cookbook-openshift3']['openshift_common_api_hostname']
  default['cookbook-openshift3']['openshift_master_embedded_etcd'] = true
  default['cookbook-openshift3']['openshift_master_etcd_port'] = '4001'
  default['cookbook-openshift3']['master_etcd_cert_prefix'] = ''
end

default['cookbook-openshift3']['ose_version'] = nil
default['cookbook-openshift3']['persistent_storage'] = []
default['cookbook-openshift3']['openshift_deployment_type'] = 'enterprise'
default['cookbook-openshift3']['ose_major_version'] = node['cookbook-openshift3']['openshift_deployment_type'] =~ /enterprise/ ? '3.4' : '1.4'
default['cookbook-openshift3']['openshift_docker_image_version'] = node['cookbook-openshift3']['openshift_deployment_type'] =~ /enterprise/ ? 'v3.4' : 'v1.4.1'
default['cookbook-openshift3']['deploy_containerized'] = false
default['cookbook-openshift3']['deploy_example'] = false
default['cookbook-openshift3']['deploy_dnsmasq'] = false
default['cookbook-openshift3']['deploy_standalone_registry'] = false
default['cookbook-openshift3']['deploy_example_db_templates'] = true
default['cookbook-openshift3']['deploy_example_image-streams'] = true
default['cookbook-openshift3']['deploy_example_quickstart-templates'] = false
default['cookbook-openshift3']['deploy_example_xpaas-streams'] = false
default['cookbook-openshift3']['deploy_example_xpaas-templates'] = false

default['cookbook-openshift3']['docker_version'] = nil
default['cookbook-openshift3']['docker_log_driver'] = 'json-file'
default['cookbook-openshift3']['docker_log_options'] = {}
default['cookbook-openshift3']['install_method'] = 'yum'
default['cookbook-openshift3']['httpd_xfer_port'] = '9999'
default['cookbook-openshift3']['core_packages'] = %w(libselinux-python wget vim-enhanced net-tools bind-utils git bash-completion bash-completion dnsmasq)
default['cookbook-openshift3']['osn_cluster_dns_domain'] = 'cluster.local'
default['cookbook-openshift3']['osn_cluster_dns_ip'] = node['ipaddress']
default['cookbook-openshift3']['enabled_firewall_rules_master'] = %w(firewall_master)
default['cookbook-openshift3']['enabled_firewall_rules_master_cluster'] = %w(firewall_master_cluster)
default['cookbook-openshift3']['enabled_firewall_rules_node'] = %w(firewall_node)
default['cookbook-openshift3']['enabled_firewall_additional_rules_node'] = []
default['cookbook-openshift3']['enabled_firewall_rules_etcd'] = %w(firewall_etcd)
default['cookbook-openshift3']['openshift_service_type'] = node['cookbook-openshift3']['openshift_deployment_type'] =~ /enterprise/ ? 'atomic-openshift' : 'origin'
default['cookbook-openshift3']['registry_persistent_volume'] = ''
default['cookbook-openshift3']['yum_repositories'] = node['cookbook-openshift3']['openshift_deployment_type'] =~ /enterprise/ ? %w() : [{ 'name' => 'centos-openshift-origin', 'baseurl' => 'http://mirror.centos.org/centos/7/paas/x86_64/openshift-origin/', 'gpgcheck' => false }]

default['cookbook-openshift3']['openshift_data_dir'] = '/var/lib/origin'
default['cookbook-openshift3']['openshift_common_base_dir'] = '/etc/origin'
default['cookbook-openshift3']['openshift_common_master_dir'] = '/etc/origin'
default['cookbook-openshift3']['openshift_common_node_dir'] = '/etc/origin'
default['cookbook-openshift3']['openshift_common_cloud_provider_dir'] = '/etc/origin'
default['cookbook-openshift3']['openshift_common_portal_net'] = '172.30.0.0/16'
default['cookbook-openshift3']['openshift_common_first_svc_ip'] = node['cookbook-openshift3']['openshift_common_portal_net'].split('/')[0].gsub(/\.0$/, '.1')
default['cookbook-openshift3']['openshift_common_default_nodeSelector'] = 'region=user'
default['cookbook-openshift3']['openshift_common_examples_base'] = '/usr/share/openshift/examples'
default['cookbook-openshift3']['openshift_common_hosted_base'] = node['cookbook-openshift3']['deploy_containerized'] == true ? '/etc/origin/hosted' : '/usr/share/openshift/hosted'
default['cookbook-openshift3']['openshift_hosted_type'] = node['cookbook-openshift3']['openshift_deployment_type'] =~ /enterprise/ ? 'enterprise' : 'origin'
default['cookbook-openshift3']['openshift_base_images'] = node['cookbook-openshift3']['openshift_deployment_type'] =~ /enterprise/ ? 'image-streams-rhel7.json' : 'image-streams-centos7.json'
default['cookbook-openshift3']['openshift_common_hostname'] = node['fqdn']
default['cookbook-openshift3']['openshift_common_ip'] = node['ipaddress']
default['cookbook-openshift3']['openshift_common_public_ip'] = node['ipaddress']
default['cookbook-openshift3']['openshift_common_admin_binary'] = node['cookbook-openshift3']['deploy_containerized'] == true ? '/usr/local/bin/oadm' : '/usr/bin/oadm'
default['cookbook-openshift3']['openshift_common_client_binary'] = node['cookbook-openshift3']['deploy_containerized'] == true ? '/usr/local/bin/oc' : '/usr/bin/oc'
default['cookbook-openshift3']['openshift_common_service_accounts'] = []
default['cookbook-openshift3']['openshift_common_service_accounts'] = [{ 'name' => 'router', 'namespace' => 'default', 'scc' => 'hostnetwork' }]
default['cookbook-openshift3']['openshift_common_service_accounts_additional'] = []
default['cookbook-openshift3']['openshift_common_use_openshift_sdn'] = true
default['cookbook-openshift3']['openshift_common_sdn_network_plugin_name'] = 'redhat/openshift-ovs-subnet'
default['cookbook-openshift3']['openshift_common_svc_names'] = ['openshift', 'openshift.default', 'openshift.default.svc', "openshift.default.svc.#{node['cookbook-openshift3']['osn_cluster_dns_domain']}", 'kubernetes', 'kubernetes.default', 'kubernetes.default.svc', "kubernetes.default.svc.#{node['cookbook-openshift3']['osn_cluster_dns_domain']}", node['cookbook-openshift3']['openshift_common_first_svc_ip']]
default['cookbook-openshift3']['openshift_common_registry_url'] = node['cookbook-openshift3']['openshift_deployment_type'] =~ /enterprise/ ? 'openshift3/ose-${component}:${version}' : 'openshift/origin-${component}:${version}'
default['cookbook-openshift3']['openshift_cloud_provider_config_dir'] = "#{node['cookbook-openshift3']['openshift_common_cloud_provider_dir']}/cloudprovider"
default['cookbook-openshift3']['openshift_docker_insecure_registry_arg'] = []
default['cookbook-openshift3']['openshift_docker_add_registry_arg'] = []
default['cookbook-openshift3']['openshift_docker_block_registry_arg'] = []
default['cookbook-openshift3']['openshift_docker_insecure_registries'] = node['cookbook-openshift3']['openshift_docker_add_registry_arg'].empty? ? [node['cookbook-openshift3']['openshift_common_portal_net']] : [node['cookbook-openshift3']['openshift_common_portal_net']] + node['cookbook-openshift3']['openshift_docker_insecure_registry_arg']
default['cookbook-openshift3']['openshift_docker_cli_image'] = node['cookbook-openshift3']['openshift_deployment_type'] =~ /enterprise/ ? 'openshift3/ose' : 'openshift/origin'
default['cookbook-openshift3']['openshift_docker_master_image'] = node['cookbook-openshift3']['openshift_deployment_type'] =~ /enterprise/ ? 'openshift3/ose' : 'openshift/origin'
default['cookbook-openshift3']['openshift_docker_node_image'] = node['cookbook-openshift3']['openshift_deployment_type'] =~ /enterprise/ ? 'openshift3/node' : 'openshift/node'
default['cookbook-openshift3']['openshift_docker_ovs_image'] = node['cookbook-openshift3']['openshift_deployment_type'] =~ /enterprise/ ? 'openshift3/openvswitch' : 'openshift/openvswitch'
default['cookbook-openshift3']['openshift_docker_etcd_image'] = 'registry.access.redhat.com/rhel7/etcd'
default['cookbook-openshift3']['openshift_master_config_dir'] = "#{node['cookbook-openshift3']['openshift_common_master_dir']}/master"
default['cookbook-openshift3']['openshift_master_bind_addr'] = '0.0.0.0'
default['cookbook-openshift3']['openshift_master_auditconfig'] = false
default['cookbook-openshift3']['openshift_master_api_port'] = '8443'
default['cookbook-openshift3']['openshift_master_console_port'] = '8443'
default['cookbook-openshift3']['openshift_master_controllers_port'] = '8444'
default['cookbook-openshift3']['openshift_master_controller_lease_ttl'] = '30'
default['cookbook-openshift3']['openshift_master_dynamic_provisioning_enabled'] = true
default['cookbook-openshift3']['openshift_master_disabled_features'] = "['Builder', 'S2IBuilder', 'WebConsole']"
default['cookbook-openshift3']['openshift_master_embedded_dns'] = true
default['cookbook-openshift3']['openshift_master_embedded_kube'] = true
default['cookbook-openshift3']['openshift_master_debug_level'] = '2'
default['cookbook-openshift3']['openshift_master_dns_port'] = node['cookbook-openshift3']['deploy_dnsmasq'] == true ? '8053' : '53'
default['cookbook-openshift3']['openshift_master_metrics_public_url'] = nil
default['cookbook-openshift3']['openshift_master_image_bulk_imported'] = 5
default['cookbook-openshift3']['openshift_master_deserialization_cache_size'] = '50000'
default['cookbook-openshift3']['openshift_master_pod_eviction_timeout'] = ''
default['cookbook-openshift3']['openshift_master_project_request_message'] = ''
default['cookbook-openshift3']['openshift_master_project_request_template'] = ''
default['cookbook-openshift3']['openshift_master_logging_public_url'] = nil
default['cookbook-openshift3']['openshift_master_router_subdomain'] = 'cloudapps.domain.local'
default['cookbook-openshift3']['openshift_master_sdn_cluster_network_cidr'] = '10.1.0.0/16'
default['cookbook-openshift3']['openshift_master_sdn_host_subnet_length'] = '8'
default['cookbook-openshift3']['openshift_master_oauth_grant_method'] = 'auto'
default['cookbook-openshift3']['openshift_master_session_max_seconds'] = '3600'
default['cookbook-openshift3']['openshift_master_session_name'] = 'ssn'
default['cookbook-openshift3']['openshift_master_session_secrets_file'] = "#{node['cookbook-openshift3']['openshift_master_config_dir']}/session-secrets.yaml"
default['cookbook-openshift3']['openshift_master_access_token_max_seconds'] = '86400'
default['cookbook-openshift3']['openshift_master_auth_token_max_seconds'] = '500'
default['cookbook-openshift3']['openshift_master_public_api_url'] = "https://#{node['cookbook-openshift3']['openshift_common_public_hostname']}:#{node['cookbook-openshift3']['openshift_master_api_port']}"
default['cookbook-openshift3']['openshift_master_api_url'] = "https://#{node['cookbook-openshift3']['openshift_common_api_hostname']}:#{node['cookbook-openshift3']['openshift_master_api_port']}"
default['cookbook-openshift3']['openshift_master_loopback_api_url'] = "https://#{node['fqdn']}:#{node['cookbook-openshift3']['openshift_master_api_port']}"
default['cookbook-openshift3']['openshift_master_console_url'] = "https://#{node['cookbook-openshift3']['openshift_common_public_hostname']}:#{node['cookbook-openshift3']['openshift_master_console_port']}/console"
default['cookbook-openshift3']['openshift_master_policy'] = "#{node['cookbook-openshift3']['openshift_master_config_dir']}/policy.json"
default['cookbook-openshift3']['openshift_master_config_file'] = "#{node['cookbook-openshift3']['openshift_master_config_dir']}/master-config.yaml"
default['cookbook-openshift3']['openshift_master_api_sysconfig'] = "/etc/sysconfig/#{node['cookbook-openshift3']['openshift_service_type']}-master-api"
default['cookbook-openshift3']['openshift_master_api_systemd'] = "/usr/lib/systemd/system/#{node['cookbook-openshift3']['openshift_service_type']}-master-api.service"
default['cookbook-openshift3']['openshift_master_controllers_sysconfig'] = "/etc/sysconfig/#{node['cookbook-openshift3']['openshift_service_type']}-master-controllers"
default['cookbook-openshift3']['openshift_master_controllers_systemd'] = "/usr/lib/systemd/system/#{node['cookbook-openshift3']['openshift_service_type']}-master-controllers.service"
default['cookbook-openshift3']['openshift_master_ca_certificate'] = { 'data_bag_name' => nil, 'data_bag_item_name' => nil, 'secret_file' => nil }
default['cookbook-openshift3']['openshift_master_named_certificates'] = %w()
default['cookbook-openshift3']['openshift_master_scheduler_conf'] = "#{node['cookbook-openshift3']['openshift_master_config_dir']}/scheduler.json"
default['cookbook-openshift3']['openshift_master_managed_names_additional'] = %w()
default['cookbook-openshift3']['openshift_node_config_dir'] = "#{node['cookbook-openshift3']['openshift_common_node_dir']}/node"
default['cookbook-openshift3']['openshift_node_config_file'] = "#{node['cookbook-openshift3']['openshift_node_config_dir']}/node-config.yaml"
default['cookbook-openshift3']['openshift_node_debug_level'] = '2'
default['cookbook-openshift3']['openshift_node_docker-storage'] = {}
default['cookbook-openshift3']['openshift_node_generated_configs_dir'] = '/var/www/html/node/generated-configs'
default['cookbook-openshift3']['openshift_node_iptables_sync_period'] = '5s'
default['cookbook-openshift3']['openshift_node_max_pod'] = '40'
default['cookbook-openshift3']['openshift_node_sdn_mtu_sdn'] = '1450'
default['cookbook-openshift3']['openshift_node_minimum_container_ttl_duration'] = '10s'
default['cookbook-openshift3']['openshift_node_maximum_dead_containers_per_container'] = '2'
default['cookbook-openshift3']['openshift_node_maximum_dead_containers'] = '100'
default['cookbook-openshift3']['openshift_node_image_gc_high_threshold'] = '90'
default['cookbook-openshift3']['openshift_node_image_gc_low_threshold'] = '80'
default['cookbook-openshift3']['openshift_node_cadvisor_port'] = nil # usually set to '4194'
default['cookbook-openshift3']['openshift_node_read_only_port'] = nil # usually set to '10255'

default['cookbook-openshift3']['openshift_hosted_manage_router'] = true
default['cookbook-openshift3']['openshift_hosted_router_selector'] = 'region=infra'
default['cookbook-openshift3']['openshift_hosted_router_namespace'] = 'default'
default['cookbook-openshift3']['openshift_hosted_router_certfile'] = "#{node['cookbook-openshift3']['openshift_master_config_dir']}/openshift-router.crt"
default['cookbook-openshift3']['openshift_hosted_router_keyfile'] = "#{node['cookbook-openshift3']['openshift_master_config_dir']}/openshift-router.key"

default['cookbook-openshift3']['openshift_hosted_manage_registry'] = true
default['cookbook-openshift3']['openshift_hosted_registry_selector'] = 'region=infra'
default['cookbook-openshift3']['openshift_hosted_registry_namespace'] = 'default'

default['cookbook-openshift3']['openshift_hosted_cluster_metrics'] = false
default['cookbook-openshift3']['openshift_hosted_metrics_secrets'] = ''
default['cookbook-openshift3']['openshift_hosted_metrics_parameters'] = {}

default['cookbook-openshift3']['erb_corsAllowedOrigins'] = ['127.0.0.1', 'localhost', node['cookbook-openshift3']['openshift_common_hostname'], node['cookbook-openshift3']['openshift_common_api_hostname'], node['cookbook-openshift3']['openshift_common_public_hostname']].uniq + node['cookbook-openshift3']['openshift_common_svc_names']

default['cookbook-openshift3']['master_generated_certs_dir'] = '/var/www/html/master/generated_certs'
default['cookbook-openshift3']['etcd_add_additional_nodes'] = false
default['cookbook-openshift3']['etcd_service_name'] = node['cookbook-openshift3']['deploy_containerized'] == true ? 'etcd_container' : 'etcd'
default['cookbook-openshift3']['etcd_remove_servers'] = []
default['cookbook-openshift3']['etcd_conf_dir'] = '/etc/etcd'
default['cookbook-openshift3']['etcd_ca_dir'] = "#{node['cookbook-openshift3']['etcd_conf_dir']}/ca"
default['cookbook-openshift3']['etcd_generated_certs_dir'] = '/var/www/html/etcd/generated_certs'
default['cookbook-openshift3']['etcd_ca_cert'] = "#{node['cookbook-openshift3']['etcd_conf_dir']}/ca.crt"
default['cookbook-openshift3']['etcd_cert_file'] = "#{node['cookbook-openshift3']['etcd_conf_dir']}/server.crt"
default['cookbook-openshift3']['etcd_cert_key'] = "#{node['cookbook-openshift3']['etcd_conf_dir']}/server.key"
default['cookbook-openshift3']['etcd_peer_file'] = "#{node['cookbook-openshift3']['etcd_conf_dir']}/peer.crt"
default['cookbook-openshift3']['etcd_peer_key'] = "#{node['cookbook-openshift3']['etcd_conf_dir']}/peer.key"
default['cookbook-openshift3']['etcd_openssl_conf'] = "#{node['cookbook-openshift3']['etcd_ca_dir']}/openssl.cnf"
default['cookbook-openshift3']['etcd_ca_name'] = 'etcd_ca'
default['cookbook-openshift3']['etcd_req_ext'] = 'etcd_v3_req'
default['cookbook-openshift3']['etcd_ca_exts_peer'] = 'etcd_v3_ca_peer'
default['cookbook-openshift3']['etcd_ca_exts_server'] = 'etcd_v3_ca_server'

default['cookbook-openshift3']['etcd_initial_cluster_state'] = 'new'
default['cookbook-openshift3']['etcd_initial_cluster_token'] = 'etcd-cluster-1'
default['cookbook-openshift3']['etcd_data_dir'] = '/var/lib/etcd/'
default['cookbook-openshift3']['etcd_default_days'] = '365'

default['cookbook-openshift3']['etcd_client_port'] = '2379'
default['cookbook-openshift3']['etcd_peer_port'] = '2380'

default['cookbook-openshift3']['docker_dns_search_option'] = %w()
